### PR TITLE
Configure CircleCI

### PR DESCRIPTION
**Pull request by mroderick**
This PR configures CircleCI for the repository.

#### How to verify

1. Observe that Circle builds the branch succesfully

### CODE REVIEW

**Comment by mroderick** [comment]
Almost there, except for Coveralls.

I can't quite figure out how to pass the coveralls repository token to the build without making it available to the world.

https://docs.coveralls.io/supported-ci-services#circle-jenkins-semaphore-and-codeship

> When using these CI’s you must either include your repo token in a .coveralls.yml file or, if you don’t want it under source control, set it in your build config like this in the “Test Commands” (CircleCI) or “Build Commands” (Semaphore) section of the project settings:
> `COVERALLS_REPO_TOKEN=asdfasdf bundle exec rspec spec`

Since the command that needs it is not a test command (and most likely refers to Circle 1.0), I don't see how this will work.

If I turn on the feature in Circle to share secrets with forks, then I might as well just put the secret into `.coveralls.yml`.

An alternative would be to switch to codecov.io, which has circle support: 

https://circleci.com/blog/making-code-coverage-easy-to-see-with-the-codecov-orb/

@fearphage you mentioned that you have some experience configuring Circle. I am too tired to figure out how to configure it so that we test in 3 node versions, only run lint in 10 and only upload coverage reports once. Could you help out here?

**Comment by mroderick** [comment]
OK. I managed to get codecov to work. @sinonjs/sinon-org-core are we happy to switch to that, when we're switching to CircleCI?

If so, then I'll rebase this branch down to a reasonable set of commits and we can merge it

**Comment by fearphage** [comment]
@mroderick You set it up in the CircleCI interface as an environment variable.

**Comment by mroderick** [comment]
> @mroderick You set it up in the CircleCI interface as an environment variable.

I did, but environment variables are not shared with builds from forks.

2019-03-10 at 09 17

**Comment by mantoni** [comment]
> but environment variables are not shared with builds from forks.

That is precisely the reason why we have to exclude SauceLabs builds in PR from forks (see https://github.com/sinonjs/sinon/issues/1013).

There are security implications with allowing environment variables in builds from forks:

> A pull request sent from a fork of the upstream repository could be manipulated to expose environment variables. The upstream repository’s maintainer would have no protection against this attack, as pull requests can be sent by anyone who forks the repository on GitHub.

**Comment by mroderick** [comment]
So, shall we just go with codecov, which works for PRs?

**Comment by mantoni** [comment]
I'm surprised coveralls doesn't have proper GitHub integration. So yeah, let's go with codecov.

File: .circleci/config.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ jobs:
     docker:
       - image: node
     environment:
-      NPM_VERSION: "6.7.0"
+      NPM_VERSION: "5.7.0"
```
Comment (mantoni) [inline .circleci/config.yml]: Why did you change to an older npm?

Comment (mroderick) [reply .circleci/config.yml]: Because latest npm failed on node 6 ... since we're only going to use node six for another month, it'll be fine :)
